nifi-nar-bundles/nifi-kafka-bundle/nifi-kafka-pubsub-processors/src/main/java/org/apache/nifi/processors/kafka/pubsub/ConsumeKafka.java

```diff
@@ -0,0 +1,213 @@
+
+    private volatile String topic;
+
+    /*
+     * Will ensure that list of PropertyDescriptors is build only once, since
+     * all other lifecycle methods are invoked multiple times.
+     */
+    static {
+        List<PropertyDescriptor> _descriptors = new ArrayList<>();
+        _descriptors.addAll(sharedDescriptors);
+        _descriptors.add(GROUP_ID);
+        _descriptors.add(AUTO_OFFSET_RESET);
+        _descriptors.add(MESSAGE_DEMARCATOR);
+        descriptors = Collections.unmodifiableList(_descriptors);
+
+        relationships = Collections.unmodifiableSet(sharedRelationships);
+    }
+
+    /**
+     *
+     */
+    @Override
+    public Set<Relationship> getRelationships() {
+        return relationships;
+    }
+
+    /**
+     * Will unsubscribe form {@link KafkaConsumer} delegating to 'super' to do
+     * the rest.
+     */
+    @Override
+    @OnStopped
+    public void close() {
+        if (this.kafkaResource != null) {
+            try {
+                this.kafkaResource.unsubscribe();
+            } finally { // in the event the above fails
+                super.close();
+            }
+        }
+    }
+
+    /**
+     *
+     */
+    @Override
+    protected List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return descriptors;
+    }
+
+    /**
+     * Will rendezvous with Kafka by performing the following:
+     * <br>
+     * - poll {@link ConsumerRecords} from {@link KafkaConsumer} in a
+     * non-blocking manner, signaling yield if no records were received from
+     * Kafka
+     * <br>
+     * - if records were received form Kafka, the are written to a newly created
+     * {@link FlowFile}'s {@link OutputStream} using a provided demarcator (see
+     * {@link #MESSAGE_DEMARCATOR}
+     */
+    @Override
+    protected void rendezvousWithKafka(ProcessContext context, ProcessSession processSession) throws ProcessException {
+        ConsumerRecords<byte[], byte[]> consumedRecords = this.kafkaResource.poll(100);
+        if (consumedRecords != null && !consumedRecords.isEmpty()) {
+            long start = System.nanoTime();
+            FlowFile flowFile = processSession.create();
+            final AtomicInteger messageCounter = new AtomicInteger();
+
+            for (final ConsumerRecord<byte[], byte[]> consumedRecord : consumedRecords) {
+                flowFile = processSession.append(flowFile, new OutputStreamCallback() {
+                    @Override
+                    public void process(final OutputStream out) throws IOException {
+                        if (messageCounter.getAndIncrement() > 0) {
+                            out.write(ConsumeKafka.this.demarcatorBytes);
+                        }
+                        out.write(consumedRecord.value());
+                    }
+                });
+            }
+
+            this.releaseFlowFile(flowFile, processSession, flowFile.getAttributes(), start, messageCounter.get() - 1);
```

Comment:
When testing in a simple case (no demarcator), receive provenance events are always showing "Received 0 Kafka messages".